contracts/extensions/version-control/ConditionalImplementControl.sol

```diff
@@ -0,0 +1,164 @@
+import { IConditionalImplementControl } from "../../interfaces/version-control/IConditionalImplementControl.sol";
+import { ErrorHandler } from "../../libraries/ErrorHandler.sol";
+import { AddressArrayUtils } from "../../libraries/AddressArrayUtils.sol";
+import { ErrOnlySelfCall } from "../../utils/CommonErrors.sol";
+
+/**
+ * @title ConditionalImplementControl
+ * @dev A contract that allows conditional version control of contract implementations.
+ */
+abstract contract ConditionalImplementControl is IConditionalImplementControl, ERC1967Upgrade {
+  using ErrorHandler for bool;
+  using AddressArrayUtils for address[];
+
+  address public immutable PROXY_STORAGE;
+  address public immutable NEW_VERSION;
+  address public immutable CURRENT_VERSION;
+
+  /**
+   * @dev Modifier that only allows self calls.
+   */
+  modifier onlySelfCall() {
+    _requireSelfCall();
+    _;
+  }
+
+  /**
+   * @dev Modifier that executes the function when conditions are met.
+   */
+  modifier whenConditionsAreMet() virtual {
+    _;
+    if (_isConditionMet()) {
+      try this.selfMigrate{ gas: _gasStipenedNoGrief() }() {} catch {}
+    }
+  }
+
+  /**
+   * @dev Modifier that only allows delegate calls from the admin proxy storage.
+   */
+  modifier onlyDelegateFromProxyStorage() virtual {
+    _requireDelegateFromProxyStorage();
+    _;
+  }
+
+  /**
+   * @dev Modifier that only allows contracts with code.
+   * @param addr The address of the contract to check.
+   */
+  modifier onlyContract(address addr) {
+    _requireHasCode(addr);
+    _;
+  }
+
+  /**
+   * @dev Constructs the ConditionalImplementControl contract.
+   * @param proxyStorage_ The address of the proxy that is allowed to delegate to this contract.
+   * @param currentVersion_ The address of the current contract implementation.
+   * @param newVersion_ The address of the new contract implementation.
+   */
+  constructor(
+    address proxyStorage_,
+    address currentVersion_,
+    address newVersion_
+  ) onlyContract(proxyStorage_) onlyContract(currentVersion_) onlyContract(newVersion_) {
+    address[] memory addrs = new address[](3);
+    addrs[0] = proxyStorage_;
+    addrs[1] = currentVersion_;
+    addrs[2] = newVersion_;
+    if (addrs.hasDuplicate()) revert AddressArrayUtils.ErrDuplicated(msg.sig);
+
+    PROXY_STORAGE = proxyStorage_;
+    NEW_VERSION = newVersion_;
+    CURRENT_VERSION = currentVersion_;
+  }
+
+  /**
+   * @dev Fallback function that forwards the call to the current or new contract implementation based on a condition.
+   */
+  fallback() external payable virtual onlyDelegateFromProxyStorage {
+    _fallback();
+  }
+
+  receive() external payable virtual onlyDelegateFromProxyStorage {
+    _fallback();
+  }
+
+  /**
+   * @dev Executes the selfMigrate function, upgrading to the new contract implementation.
+   */
+  function selfMigrate() external onlyDelegateFromProxyStorage onlySelfCall {
```

Comment:
I think `selfUpgrade` is a better name, since this will upgrade the contract to a new one. Migration sounds more related to data migration.

```suggestion
  function selfUpgrade() external onlyDelegateFromProxyStorage onlySelfCall {
```